File: android/app/src/main/java/com/kordar/ghostchat/core/security/BiometricAuthService.kt
package com.kordar.ghostchat.core.security

import com.kordar.ghostchat.models.AutoLockTimeout
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * PIN + biometric authentication service. Mirrors iOS `BiometricAuthService`:
 *
 * - SHA-256(salt || pin) hash (identical to iOS)
 * - main PIN + optional decoy PIN
 * - 10 failed attempts → panic wipe (invokes `Config.onWipe`)
 * - biometric gate (BiometricPrompt launcher is injected so the service itself
 *   stays free of Android UI APIs and testable)
 */
class BiometricAuthService(
    private val keystore: KeystoreServicing,
    private val biometricLauncher: BiometricLauncher = BiometricLauncher.Unavailable,
    private var config: Config = Config()
) {

    sealed class AuthResult {
        data object Authenticated      : AuthResult()
        data object AuthenticatedDecoy : AuthResult()
        data object Invalid            : AuthResult()
        data object Wiped              : AuthResult()
    }

    data class Config(
        val failureLimit: Int = 10,
        val onWipe: () -> Unit = {}
    )

    object Keys {
        const val MAIN_PIN      = "auth.pin.main"
        const val DECOY_PIN     = "auth.pin.decoy"
        const val FAILURE_COUNT = "auth.fail.count"
        const val BIOMETRIC     = "auth.biometric.enabled"
        const val AUTO_LOCK_SEC = "auth.autolock.seconds"
    }

    /** PIN entry validation — same shape as iOS. */
    enum class PINError { INVALID_FORMAT }

    // MARK: - PIN management

    fun setMainPIN(pin: String) {
        validate(pin)
        val hash = PINHash.create(pin)
        keystore.set(Keys.MAIN_PIN, hash.encoded())
    }

    fun setDecoyPIN(pin: String) {
        validate(pin)
        val hash = PINHash.create(pin)
        keystore.set(Keys.DECOY_PIN, hash.encoded())
    }

    fun clearPINs() {
        keystore.delete(Keys.MAIN_PIN)
        keystore.delete(Keys.DECOY_PIN)
        resetFailureCount()
    }

    fun hasMainPIN(): Boolean = keystore.get(Keys.MAIN_PIN) != null

    // MARK: - Authenticate

    fun authenticate(pin: String): AuthResult {
        keystore.get(Keys.MAIN_PIN)?.let { raw ->
            val hash = PINHash.fromEncoded(raw)
            if (hash != null && hash.verify(pin)) {
                resetFailureCount()
                return AuthResult.Authenticated
            }
        }
        keystore.get(Keys.DECOY_PIN)?.let { raw ->
            val hash = PINHash.fromEncoded(raw)
            if (hash != null && hash.verify(pin)) {
                resetFailureCount()
                return AuthResult.AuthenticatedDecoy
            }
        }
        val count = incrementFailureCount()
        if (count >= config.failureLimit) {
            clearPINs()
            config.onWipe()
            return AuthResult.Wiped
        }
        return AuthResult.Invalid
    }

    /** Suspends until biometric check completes — delegates to the injected launcher. */
    suspend fun authenticateBiometric(reason: String): Boolean =
        biometricLauncher.prompt(reason)

    // MARK: - Failure counter

    fun failureCount(): Int =
        keystore.get(Keys.FAILURE_COUNT)?.let { String(it, StandardCharsets.UTF_8).toIntOrNull() } ?: 0

    private fun incrementFailureCount(): Int {
        val next = failureCount() + 1
        keystore.set(Keys.FAILURE_COUNT, next.toString().toByteArray(StandardCharsets.UTF_8))
        return next
    }

    private fun resetFailureCount() {
        keystore.set(Keys.FAILURE_COUNT, "0".toByteArray(StandardCharsets.UTF_8))
    }

    // MARK: - Biometric toggle

    var biometricEnabled: Boolean
        get() = (keystore.get(Keys.BIOMETRIC)?.firstOrNull() == 0x01.toByte())
        set(value) = keystore.set(Keys.BIOMETRIC, byteArrayOf(if (value) 0x01 else 0x00))

    // MARK: - Auto-lock

    var autoLockTimeout: AutoLockTimeout
        get() {
            val raw = keystore.get(Keys.AUTO_LOCK_SEC) ?: return AutoLockTimeout.ONE_MINUTE
            val sec = String(raw, StandardCharsets.UTF_8).toIntOrNull() ?: return AutoLockTimeout.ONE_MINUTE
            return AutoLockTimeout.fromSeconds(sec)
        }
        set(value) = keystore.set(Keys.AUTO_LOCK_SEC, value.seconds.toString().toByteArray(StandardCharsets.UTF_8))

    // MARK: - Private

    private fun validate(pin: String) {
        if (pin.length !in 4..6 || !pin.all { it.isDigit() }) {
            throw IllegalArgumentException(PINError.INVALID_FORMAT.name)
        }
    }
}

/** Salted SHA-256 digest of a PIN. Encoded form is `JSON { salt: base16, hash: base16 }`. */
internal data class PINHash(val salt: ByteArray, val hash: ByteArray) {

    fun verify(pin: String): Boolean = compute(pin, salt).contentEquals(hash)

    fun encoded(): ByteArray = Json.encodeToString(Encoded.serializer(),
        Encoded(salt = salt.toHex(), hash = hash.toHex())
    ).toByteArray(StandardCharsets.UTF_8)

    @Serializable
    private data class Encoded(val salt: String, val hash: String)

    companion object {
        fun create(pin: String): PINHash {
            val salt = randomSalt()
            return PINHash(salt, compute(pin, salt))
        }

        fun fromEncoded(data: ByteArray): PINHash? = runCatching {
            val raw = String(data, StandardCharsets.UTF_8)
            val encoded = Json.decodeFromString(Encoded.serializer(), raw)
            PINHash(salt = encoded.salt.fromHex(), hash = encoded.hash.fromHex())
        }.getOrNull()

        fun compute(pin: String, salt: ByteArray): ByteArray {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update(salt)
            digest.update(pin.toByteArray(StandardCharsets.UTF_8))
            return digest.digest()
        }

        fun randomSalt(length: Int = 32): ByteArray =
            ByteArray(length).also { SecureRandom().nextBytes(it) }

        private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

        private fun String.fromHex(): ByteArray {
            check(length % 2 == 0) { "odd hex length" }
            return chunked(2).map { it.toInt(16).toByte() }.toByteArray()
        }
    }
}

/** Abstraction over BiometricPrompt so the service is testable without an Activity. */
fun interface BiometricLauncher {
    suspend fun prompt(reason: String): Boolean

    object Unavailable : BiometricLauncher {
        override suspend fun prompt(reason: String): Boolean = false
    }
}

File: ios/GhostChat/Core/Security/BiometricAuthService.swift
import CryptoKit
import Foundation
import LocalAuthentication

/// Salted SHA-256 hash of a user PIN.
struct PINHash: Codable, Equatable {
    let salt: Data
    let hash: Data

    static func make(pin: String) -> PINHash {
        let salt = randomSalt()
        return PINHash(salt: salt, hash: Self.compute(pin: pin, salt: salt))
    }

    func verify(pin: String) -> Bool {
        Self.compute(pin: pin, salt: salt) == hash
    }

    static func compute(pin: String, salt: Data) -> Data {
        var input = Data()
        input.append(salt)
        input.append(contentsOf: pin.utf8)
        return Data(SHA256.hash(data: input))
    }

    static func randomSalt(length: Int = 32) -> Data {
        var bytes = Data(count: length)
        let status = bytes.withUnsafeMutableBytes { buf in
            SecRandomCopyBytes(kSecRandomDefault, length, buf.baseAddress!)
        }
        precondition(status == errSecSuccess, "SecRandomCopyBytes failed")
        return bytes
    }
}

final class BiometricAuthService {

    enum AuthResult: Equatable {
        case authenticated
        case authenticatedAsDecoy
        case invalid
        case wiped
    }

    struct Keys {
        static let mainPIN       = "auth.pin.main"
        static let decoyPIN      = "auth.pin.decoy"
        static let failureCount  = "auth.fail.count"
        static let biometric     = "auth.biometric.enabled"
        static let autoLockSec   = "auth.autolock.seconds"
    }

    struct Config {
        var failureLimit: Int = 10
        var onWipe: () -> Void = {}
    }

    private let keychain: KeychainServicing
    private var config: Config

    init(keychain: KeychainServicing, config: Config = .init()) {
        self.keychain = keychain
        self.config = config
    }

    // MARK: - PIN management

    func setMainPIN(_ pin: String) throws {
        try validate(pin: pin)
        let hash = PINHash.make(pin: pin)
        try keychain.set(try JSONEncoder().encode(hash), for: Keys.mainPIN)
    }

    func setDecoyPIN(_ pin: String) throws {
        try validate(pin: pin)
        let hash = PINHash.make(pin: pin)
        try keychain.set(try JSONEncoder().encode(hash), for: Keys.decoyPIN)
    }

    func clearPINs() throws {
        try keychain.delete(Keys.mainPIN)
        try keychain.delete(Keys.decoyPIN)
        try resetFailureCount()
    }

    var hasMainPIN: Bool {
        (try? keychain.get(Keys.mainPIN)) != nil
    }

    // MARK: - Authenticate

    func authenticate(pin: String) throws -> AuthResult {
        if let mainData = try keychain.get(Keys.mainPIN),
           let main = try? JSONDecoder().decode(PINHash.self, from: mainData),
           main.verify(pin: pin) {
            try resetFailureCount()
            return .authenticated
        }
        if let decoyData = try keychain.get(Keys.decoyPIN),
           let decoy = try? JSONDecoder().decode(PINHash.self, from: decoyData),
           decoy.verify(pin: pin) {
            try resetFailureCount()
            return .authenticatedAsDecoy
        }
        let count = try incrementFailureCount()
        if count >= config.failureLimit {
            try clearPINs()
            config.onWipe()
            return .wiped
        }
        return .invalid
    }

    func authenticateBiometric(reason: String) async -> Bool {
        let context = LAContext()
        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            return false
        }
        return await withCheckedContinuation { cont in
            context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, _ in
                cont.resume(returning: success)
            }
        }
    }

    // MARK: - Failure counter

    func failureCount() throws -> Int {
        let data = try keychain.get(Keys.failureCount)
        guard let bytes = data,
              let str = String(data: bytes, encoding: .utf8),
              let n = Int(str) else { return 0 }
        return n
    }

    private func incrementFailureCount() throws -> Int {
        let n = (try failureCount()) + 1
        try keychain.set(Data("\(n)".utf8), for: Keys.failureCount)
        return n
    }

    private func resetFailureCount() throws {
        try keychain.set(Data("0".utf8), for: Keys.failureCount)
    }

    // MARK: - Biometric toggle

    var biometricEnabled: Bool {
        get {
            guard let data = try? keychain.get(Keys.biometric) else { return false }
            return data == Data([0x01])
        }
        set {
            try? keychain.set(Data([newValue ? 0x01 : 0x00]), for: Keys.biometric)
        }
    }

    // MARK: - Auto-lock

    var autoLockTimeout: AutoLockTimeout {
        get {
            guard let data = try? keychain.get(Keys.autoLockSec),
                  let str = String(data: data, encoding: .utf8),
                  let seconds = Int(str),
                  let value = AutoLockTimeout(rawValue: seconds) else {
                return .oneMinute
            }
            return value
        }
        set {
            try? keychain.set(Data("\(newValue.rawValue)".utf8), for: Keys.autoLockSec)
        }
    }

    // MARK: - Validation

    private func validate(pin: String) throws {
        guard pin.count >= 4, pin.count <= 6, pin.allSatisfy({ $0.isNumber }) else {
            throw PINError.invalidFormat
        }
    }

    enum PINError: Swift.Error, Equatable {
        case invalidFormat
    }
}
